lib/src/front_end/ast_node_visitor.dart

```diff
@@ -425,37 +426,132 @@ class AstNodeVisitor extends ThrowingAstVisitor<void>
 
   @override
   void visitForStatement(ForStatement node) {
-    throw UnimplementedError();
+    token(node.forKeyword);
+    writer.split();
+    var forKeyword = writer.pop();
+
+    // Treat the for loop parts sort of as an argument list where each clause
+    // is a separate argument. This means that when they split, they split like:
+    //
+    // ```
+    // for (
+    //   initializerClause;
+    //   conditionClause;
+    //   incrementClause
+    // ) {
+    //   body;
+    // }
+    // ```
+    var partsList =
+        DelimitedListBuilder(this, const ListStyle(commas: Commas.none));
+    partsList.leftBracket(node.leftParenthesis);
+
+    // Edge case: A totally empty for loop is formatted just as `(;;)` with
+    // no splits or spaces anywhere.
+    var forLoopParts = node.forLoopParts;
+    if (forLoopParts is ForPartsWithExpression &&
+        forLoopParts.initialization == null &&
+        forLoopParts.leftSeparator.precedingComments == null &&
+        forLoopParts.condition == null &&
+        forLoopParts.rightSeparator.precedingComments == null &&
+        forLoopParts.updaters.isEmpty &&
+        node.rightParenthesis.precedingComments == null) {
+      token(forLoopParts.leftSeparator);
+      token(forLoopParts.rightSeparator);
+    } else {
+      switch (forLoopParts) {
+        // ForParts is the base class for all three-clause "C-style" for loops.
+        // There are three subclasses based on whether the first clause is a
+        // declaration, expression, or pattern. The other clauses are the same
+        // and are handled here.
+        case ForParts():
+          // The initializer clause depends on which subclass we have.
+          switch (forLoopParts) {
+            case ForPartsWithDeclarations():
+              partsList.addCommentsBefore(forLoopParts.variables.beginToken);
+              visit(forLoopParts.variables);
+
+            case ForPartsWithExpression():
+              if (forLoopParts.initialization case var initializer?) {
+                partsList.addCommentsBefore(initializer.beginToken);
+                visit(initializer);
+              } else {
+                partsList.addCommentsBefore(forLoopParts.leftSeparator);
+              }
```

Comment:
[optional] I'm less sure how I feel about this one. We can't get rid of `forLoopParts` entirely, and adding more pattern variables is maybe adding complexity instead of taking it away? I have been finding the reduced scope of pattern variables helpful for readability though, so I like to see what it looks like and I often appreciate the shorter or eliminated chains of `.` properties. The formatting throws me off a bit with this one.

```dart
            case ForPartsWithExpression(
                :var initialization,
                :var leftSeparator
              ):
              if (initialization case var initializer?) {
                partsList.addCommentsBefore(initializer.beginToken);
                visit(initializer);
              } else {
                partsList.addCommentsBefore(leftSeparator);
              }
```

Also, my gut reaction is to want to flatten the conditional into the switch. I don't know if that impacts perf? It does make the formatting nice again.

```dart
            case ForPartsWithExpression(:var initialization?):
              partsList.addCommentsBefore(initialization.beginToken);
              visit(initialization);

            case ForPartsWithExpression(:var leftSeparator):
              partsList.addCommentsBefore(leftSeparator);
```

Reply:
I think case, I kind of like nesting the if inside the case because I think it makes it a little clearer that we are handling all ForPartsWithExpression objects in that case.